### ISSUE TITLE
Training Loss Bug

### ISSUE THREAD

**Issue by coldsakrua** [georgeNakayama/AIpparel-Code#12]
Hey, when I try to full parameters fine-tune AIpparel based on V2 , I find some confused outputs.
The model is successfully loaded and training starts, however, all kinds of loss seem to be wrong.
Here is my log
```
[trainers.trainer][INFO] - Step: [1]	Time  0.446 ( 0.808)	Data  0.000 ( 0.041)	Total Loss 0.6368 (0.5791)	CE Loss 0.6250 (0.5553)	Edge Loss 0.1078 (0.2388)	image Total Loss 0.7800 (0.6163)	.......

[trainers.trainer][INFO] - Step: [2]	Time  1.192 ( 0.512)	Data  0.000 ( 0.000)	Total Loss 8.0480 (7.9632)	CE Loss 7.6875 (7.7508)	Edge Loss 3.6046 (2.1403)	image Total Loss 7.7803 (7.9115)	.......

[trainers.trainer][INFO] - Step: [3]	Time  0.572 ( 0.537)	Data  0.000 ( 0.000)	Total Loss 8.0636 (7.8924)	CE Loss 7.9375 (7.7008)	Edge Loss 1.1050 (1.9517)	image Total Loss 7.5716 (7.8885)	.......
```

Why does loss rise since step 2 ？I think it should decline from the beginning. 
Here is my train.yaml
```
_target_: trainers.trainer.Trainer
lr: 0.0006
beta1: .9
beta2: 0.95
grad_accumulation_steps: 1
batch_size: 2
num_steps: 3
save_freq: 1
```

Another confused point is that the size of the saved model is 13.89GB , which size is half of given pretrained model. What's more, the type of saved model is .pt when given is .pth. Could you explain the reason? Thanks a lot !

**Comment by georgeNakayama** [comment]
Hi, 

The loss spikes probably because the current model's state is far from convergent given your new data. Have you shifted the specifications using the shift_specs.py? The logging in the first step is basically evaluation loss before having updated the model weights. That's why the loss is lower comparatively. The loss should go down quickly as training progresses. 

For the checkpoint size, currently the logic saves the deepspeed partitioned model weights. To get the model weight used by Torch, you need to run the conversion script [here](https://github.com/georgeNakayama/AIpparel-Code/blob/master/trainers/convert_zero_to_torch.py#L474). See the logics [here](https://github.com/georgeNakayama/AIpparel-Code/blob/master/trainers/trainer.py#L577-L588) for the details. 

Let me know if there's still discrepancies!

**Comment by coldsakrua** [comment]
Hi, thanks for answering.
The conversion for the .pt checkpoint is useful ! 
But there is something wrong when resuming training. When I add the codes in run.py
```
### run.py
model.resize_token_embeddings(len(tokenizer))

###
a=torch.load(cfg.pre_trained, map_location='cpu') if cfg.pre_trained else None
    if a is not None:
        if master_process:
            log.info(f"####### Loading pre-trained model from {cfg.pre_trained}")
        model.load_state_dict(a, strict=False)
### 

total_params = sum(p.numel() for p in model.parameters())
trainable_params = sum(p.numel() for p in model.parameters() if p.requires_grad)

```
The loss declines from the begining.(I think loss declines absolutely right for finetuning  )

```
Step: [1]	Time  0.567 ( 0.784)	Data  0.000 ( 0.034)	Total Loss 0.6270 (0.5790)	CE Loss 0.6172 (0.5620)	Edge Loss 0.1075 (0.1691) 	......

Step: [2]	Time  1.138 ( 0.555)	Data  0.000 ( 0.000)	Total Loss 0.5892 (0.5824)	CE Loss 0.5781 (0.5515)	Edge Loss 0.0915 (0.3111)	......

Step: [3]	Time  0.765 ( 0.602)	Data  0.000 ( 0.000)	Total Loss 0.5365 (0.5675)	CE Loss 0.5234 (0.5561)	Edge Loss 0.1113 (0.1137)	......
```
In my opinion, the bug may happen here
```
### trainer.py line 182
self._start_experiment(in_config, resume, from_start)
```
It works when evaluating, but may change the parameters in pre-trained model, which means I trained an initialized model before.

Would you please try to resume model ? If you get similar output like loss spike and loss declines from spike (like training a new model rather than finetuning pre-trained model), there may be a bug for finetuning :). 


P.S here is origin loss changes trend
```
Step: [1]	Time  0.582 ( 0.739)	Data  0.000 ( 0.038)	Total Loss 0.6256 (0.5892)
Step: [2]	Time  1.011 ( 0.551)	Data  0.000 ( 0.000)	Total Loss 16.0141 (14.8705)
Step: [3]	Time  0.784 ( 0.614)	Data  0.000 ( 0.000)	Total Loss 13.6604 (14.6902)
Step: [4]	Time  1.146 ( 0.572)	Data  0.000 ( 0.000)	Total Loss 12.6512 (12.6179)
Step: [5]	Time  1.222 ( 0.614)	Data  0.000 ( 0.000)	Total Loss 9.3832 (12.7493)
Step: [6]	Time  1.318 ( 0.588)	Data  0.000 ( 0.000)	Total Loss 8.9957 (9.6626)
...
...
Step: [3999]	Time  1.944 ( 0.608)	Data  0.000 ( 0.000)	Total Loss 1.1629 (1.0996)
Step: [4000]	Time  0.996 ( 0.603)	Data  0.000 ( 0.000)	Total Loss 1.1769 (1.0815)

```
Just like initialize a new model and train from step2 , don't you think so? :)

**Comment by georgeNakayama** [comment]
I see! Then it might actually be a bug! I'm glad that you figured it out though. So do you think we need to load the model before initializing deepspeed?

**Comment by coldsakrua** [comment]
Yeah. Honestly I'm not good at using deepspeed, but there might be something wrong when initializing deepspeed. I fix it by loading model using pytorch before initializing deepspped, but there might be other way to solve it :)

**Comment by georgeNakayama** [comment]
Thanks! I can fix it. Or if you would like to make a pull request I'm more than happy to incorporate the changes!

**Comment by yangmingweiwtu** [comment]
> Hi, thanks for answering. The conversion for the .pt checkpoint is useful ! But there is something wrong when resuming training. When I add the codes in run.py
> 
> ```
> ### run.py
> model.resize_token_embeddings(len(tokenizer))
> 
> ###
> a=torch.load(cfg.pre_trained, map_location='cpu') if cfg.pre_trained else None
>     if a is not None:
>         if master_process:
>             log.info(f"####### Loading pre-trained model from {cfg.pre_trained}")
>         model.load_state_dict(a, strict=False)
> ### 
> 
> total_params = sum(p.numel() for p in model.parameters())
> trainable_params = sum(p.numel() for p in model.parameters() if p.requires_grad)
> ```
> 
> The loss declines from the begining.(I think loss declines absolutely right for finetuning )
> 
> ```
> Step: [1]	Time  0.567 ( 0.784)	Data  0.000 ( 0.034)	Total Loss 0.6270 (0.5790)	CE Loss 0.6172 (0.5620)	Edge Loss 0.1075 (0.1691) 	......
> 
> Step: [2]	Time  1.138 ( 0.555)	Data  0.000 ( 0.000)	Total Loss 0.5892 (0.5824)	CE Loss 0.5781 (0.5515)	Edge Loss 0.0915 (0.3111)	......
> 
> Step: [3]	Time  0.765 ( 0.602)	Data  0.000 ( 0.000)	Total Loss 0.5365 (0.5675)	CE Loss 0.5234 (0.5561)	Edge Loss 0.1113 (0.1137)	......
> ```
> 
> In my opinion, the bug may happen here
> 
> ```
> ### trainer.py line 182
> self._start_experiment(in_config, resume, from_start)
> ```
> 
> It works when evaluating, but may change the parameters in pre-trained model, which means I trained an initialized model before.
> 
> Would you please try to resume model ? If you get similar output like loss spike and loss declines from spike (like training a new model rather than finetuning pre-trained model), there may be a bug for finetuning :).
> 
> P.S here is origin loss changes trend
> 
> ```
> Step: [1]	Time  0.582 ( 0.739)	Data  0.000 ( 0.038)	Total Loss 0.6256 (0.5892)
> Step: [2]	Time  1.011 ( 0.551)	Data  0.000 ( 0.000)	Total Loss 16.0141 (14.8705)
> Step: [3]	Time  0.784 ( 0.614)	Data  0.000 ( 0.000)	Total Loss 13.6604 (14.6902)
> Step: [4]	Time  1.146 ( 0.572)	Data  0.000 ( 0.000)	Total Loss 12.6512 (12.6179)
> Step: [5]	Time  1.222 ( 0.614)	Data  0.000 ( 0.000)	Total Loss 9.3832 (12.7493)
> Step: [6]	Time  1.318 ( 0.588)	Data  0.000 ( 0.000)	Total Loss 8.9957 (9.6626)
> ...
> ...
> Step: [3999]	Time  1.944 ( 0.608)	Data  0.000 ( 0.000)	Total Loss 1.1629 (1.0996)
> Step: [4000]	Time  0.996 ( 0.603)	Data  0.000 ( 0.000)	Total Loss 1.1769 (1.0815)
> ```
> 
> Just like initialize a new model and train from step2 , don't you think so? :)
Hi, do you mean that the author retrained the model instead of fine-tuning due to a code bug? However, the author achieved better performance with this method. Can the author achieve the same results after fixing the bug?
